video/out/vo_sixel.c

```diff
@@ -0,0 +1,483 @@
+
+    /* compare color difference between current
+     * palette and previous one */
+    for (i = 0; i < palette_colors; i++) {
+        r += palette[i * 3 + 0];
+        g += palette[i * 3 + 1];
+        b += palette[i * 3 + 2];
+    }
+    score = (r - average_r) * (r - average_r)
+          + (g - average_g) * (g - average_g)
+          + (b - average_b) * (b - average_b);
+    if (score > priv->opts->threshold * palette_colors
+                             * palette_colors)
+        goto detected;
+
+    return 0;
+
+detected:
+    previous_histgram_colors = histgram_colors;
+    average_r = r;
+    average_g = g;
+    average_b = b;
+    return 1;
+}
+
+static SIXELSTATUS prepare_static_palette(struct vo* vo)
+{
+    struct priv* priv = vo->priv;
+    if (priv->dither) {
+        sixel_dither_set_body_only(priv->dither, 1);
+    } else {
+        priv->dither = sixel_dither_get(BUILTIN_XTERM256);
+        if (priv->dither == NULL)
+            return SIXEL_FALSE;
+        sixel_dither_set_diffusion_type(priv->dither, priv->opts->diffuse);
+    }
+    return SIXEL_OK;
+}
+
+static SIXELSTATUS prepare_dynamic_palette(struct vo *vo)
+{
+    SIXELSTATUS status = SIXEL_FALSE;
+    struct priv *priv = vo->priv;
+
+    /* create histgram and construct color palette
+     * with median cut algorithm. */
+    status = sixel_dither_initialize(priv->testdither, priv->buffer,
+                                     priv->opts->width, priv->opts->height, 3,
+                                     LARGE_NORM, REP_CENTER_BOX,
+                                     QUALITY_LOW);
+    if (SIXEL_FAILED(status))
+        return status;
+
+    if (detect_scene_change(vo)) {
+        if (priv->dither)
+            sixel_dither_unref(priv->dither);
+
+        priv->dither = priv->testdither;
+        status = sixel_dither_new(&priv->testdither, priv->opts->reqcolors, NULL);
+
+        if (SIXEL_FAILED(status))
+            return status;
+        sixel_dither_set_diffusion_type(priv->dither, priv->opts->diffuse);
+    } else {
+        sixel_dither_set_body_only(priv->dither, 1);
+    }
+
+    return status;
+}
+
+static int resize(struct vo *vo)
+{
+    struct priv *priv = vo->priv;
+
+    talloc_free(priv->buffer);
+
+    if (priv->dither) {
+        sixel_dither_unref(priv->dither);
+        priv->dither = NULL;
+    }
+
+    if (priv->testdither) {
+        sixel_dither_unref(priv->testdither);
+        priv->testdither = NULL;
+    }
+    SIXELSTATUS status = sixel_dither_new(&priv->testdither, priv->opts->reqcolors, NULL);
+
+    if (SIXEL_FAILED(status)) {
+        return status;
+    }
+    priv->buffer =
+        talloc_array(NULL, uint8_t, depth * priv->opts->width * priv->opts->height);
+
+    return 0;
+}
+
+static int reconfig(struct vo *vo, struct mp_image_params *params)
+{
+    struct priv *priv = vo->priv;
+    priv->image_height = params->h;
+    priv->image_width  = params->w;
+    priv->image_format = params->imgfmt;
+    scroll_on_demand(vo);
```

Comment:
What happens if we remove the `scroll_on_demand(vo)` call? maybe we can just get rid of it?

Reply:
It does work without scroll_on_demand,  I've modified the function to only store the offset location so that people may be able to use `--vo-sixel-offset-top` and `--vo-sixel-offset-left` to set the offset from topleft position. In the future, might be able to use `TIOCGWINSZ` ioctl to also get the resolution in pixels of the terminal window and choose a better default resolution for the image, but for the starting simple implementation just setting the offsets only and rendering based on user specified dimensions.